mergify_engine/worker.py

```diff
@@ -0,0 +1,366 @@
+                    exc_info=True,
+                )
+                return
+            except Exception:
+                logger.error("fail to process pull request", exc_info=True)
+
+        message_ids_to_delete = list(
+            itertools.chain.from_iterable(message_ids.values())
+        )
+        score = utils.utcnow().timestamp()
+        await self._redis.eval(
+            self.ATOMIC_CLEAN_STREAM_SCRIPT,
+            keys=[stream_name.encode()],
+            args=[score, len(message_ids_to_delete)] + message_ids_to_delete,
+        )
+
+    # NOTE(sileht): If the stream still have messages, we update the score to reschedule the
+    # pull later
+    ATOMIC_CLEAN_STREAM_SCRIPT = """
+local stream_name = KEYS[1]
+local score = ARGV[1]
+local message_ids_len =  tonumber(ARGV[2])
+local message_ids = {unpack(ARGV, 3, message_ids_len + 3)}
+
+if message_ids_len > 0 then
+    redis.call("XDEL", stream_name, unpack(message_ids))
+end
+
+local len = tonumber(redis.call('XLEN', stream_name))
+if len == 0 then
+    redis.call("ZREM", "streams", stream_name)
+    redis.call("DEL", stream_name)
+else
+    redis.call("ZADD", "streams", score, stream_name)
+end
+"""
+
+
+class Worker:
+    def __init__(self, worker_count=config.STREAM_WORKERS, idle_sleep_time=0.42):
+        self._idle_sleep_time = idle_sleep_time
+        self._worker_count = worker_count
+        self._worker_tasks = []
+        self._stream_processor = None
+
+        self._loop = asyncio.get_running_loop()
+        self._running = asyncio.Event()
+
+    async def stream_worker_task(self, worker_id):
+        # NOTE(sileht): This task must never fail, we don't want to write code to
+        # reap/clean/respawn them
+        while self._running.is_set():
+            try:
+
+                async with self._stream_processor.next() as stream_name:
+                    if stream_name:
+                        LOG.debug("worker %d take stream: %s", worker_id, stream_name)
+                        try:
+                            await self._stream_processor.consume(stream_name)
+                        finally:
+                            LOG.debug(
+                                "worker %d release stream: %s", worker_id, stream_name,
+                            )
+                    else:
+                        await asyncio.sleep(self._idle_sleep_time)
+            except Exception:
+                LOG.error("worker %d fail", worker_id, exc_info=True)
+                await asyncio.sleep(self._idle_sleep_time)
+
+        LOG.info("worker %d exited", worker_id)
+
+    async def _run(self):
+        self._running.set()
+
+        self._stream_processor = StreamProcessor(self._worker_count)
+        await self._stream_processor.connect()
+
+        for i in range(self._worker_count):
+            self._worker_tasks.append(asyncio.create_task(self.stream_worker_task(i)))
+
+        LOG.info("%d workers spawned", self._worker_count)
+
+    async def _shutdown(self):
```

Comment:
how does that play with heroku btw?

Reply:
Heroku send SIGTERM and start the shutdown, I don't known you many times it waits before sending SIGKILL.
Anyways, if everything is killed in the middle, it's not a big deal, just the ongoing pull will be reevaluated, just like Celery